pkg/handlers/generic/mutation/controlplanevirtualip/inject_test.go

```diff
@@ -268,3 +275,247 @@ spec:
+	t.Parallel()
+	tests := []struct {
+		name          string
+		obj           *controlplanev1.KubeadmControlPlaneTemplate
+		files         []bootstrapv1.File
+		expectedFiles []bootstrapv1.File
+	}{
+		{
+			name: "should merge files",
+			obj: &controlplanev1.KubeadmControlPlaneTemplate{
+				Spec: controlplanev1.KubeadmControlPlaneTemplateSpec{
+					Template: controlplanev1.KubeadmControlPlaneTemplateResource{
+						Spec: controlplanev1.KubeadmControlPlaneTemplateResourceSpec{
+							KubeadmConfigSpec: bootstrapv1.KubeadmConfigSpec{
+								Files: []bootstrapv1.File{
+									{
+										Path:    "file-1",
+										Content: "old",
+									},
+									{
+										Path:    "file-2",
+										Content: "old",
+									},
+									{
+										Path:    "file-3",
+										Content: "old",
+									},
+									{
+										Path:    "file-4",
+										Content: "old",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			files: []bootstrapv1.File{
+				{
+					Path:    "file-1",
+					Content: "new",
+				},
+				{
+					Path:    "file-4",
+					Content: "new",
+				},
+				{
+					Path:    "file-5",
+					Content: "new",
+				},
+			},
+			expectedFiles: []bootstrapv1.File{
+				{
+					Path:    "file-1",
+					Content: "new",
+				},
+				{
+					Path:    "file-2",
+					Content: "old",
+				},
+				{
+					Path:    "file-3",
+					Content: "old",
+				},
+				{
+					Path:    "file-4",
+					Content: "new",
+				},
+				{
+					Path:    "file-5",
+					Content: "new",
+				},
+			},
+		},
+		{
+			name: "should add a new file",
+			obj: &controlplanev1.KubeadmControlPlaneTemplate{
+				Spec: controlplanev1.KubeadmControlPlaneTemplateSpec{
+					Template: controlplanev1.KubeadmControlPlaneTemplateResource{
+						Spec: controlplanev1.KubeadmControlPlaneTemplateResourceSpec{
+							KubeadmConfigSpec: bootstrapv1.KubeadmConfigSpec{
+								Files: []bootstrapv1.File{
+									{
+										Path:    "file-1",
+										Content: "old",
+									},
+									{
+										Path:    "file-2",
+										Content: "old",
+									},
+									{
+										Path:    "file-3",
+										Content: "old",
+									},
+									{
+										Path:    "file-4",
+										Content: "old",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			files: []bootstrapv1.File{
+				{
+					Path:    "file-5",
+					Content: "new",
+				},
+			},
+			expectedFiles: []bootstrapv1.File{
+				{
+					Path:    "file-1",
+					Content: "old",
+				},
+				{
+					Path:    "file-2",
+					Content: "old",
+				},
+				{
+					Path:    "file-3",
+					Content: "old",
+				},
+				{
+					Path:    "file-4",
+					Content: "old",
+				},
+				{
+					Path:    "file-5",
+					Content: "new",
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			mergeFiles(tt.obj, tt.files...)
+			assert.Equal(t, tt.expectedFiles, tt.obj.Spec.Template.Spec.KubeadmConfigSpec.Files)
```

Comment:
Consider using ElementsMatch so test doesn't rely on order
```go
assert.ElementsMatch(
				t,
				tt.expectedFiles,
				tt.obj.Spec.Template.Spec.KubeadmConfigSpec.Files,
			)
```

Reply:
We do want to preserve the order, please see https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/1062#discussion_r1989907485